ofrak_core/ofrak/gui/script_builder.py

```diff
@@ -0,0 +1,280 @@
+
+from ofrak.resource import Resource
+
+
+class SelectableAttributesError(Exception):
+    """
+    Prompt the user for an attribute to select with
+    """
+
+
+class ActionType(IntEnum):
+    UNPACK = 0
+    MOD = 1
+    PACK = 2
+    UNDEF = 3
+
+
+@dataclass
+class ScriptAction:
+    """
+    Encapsulates the structure of a single action within the script, which consists of the string
+    representation of the code for that action and the action's type.
+    """
+
+    action_type: ActionType
+    action: str
+
+
+class ScriptSession:
+    """
+    A script, consisting of an ordered sequence of script actions and a mapping between resources
+    and their autogenerated variable names.
+    """
+
+    hashed_actions: OrderedDict[int, ScriptAction] = OrderedDict()
+    variable_mapping: Dict[bytes, str] = {}
+    actions_counter: int = 0
+    boilerplate_header: str = r"""
+    from ofrak import *
+    from ofrak.core import *
+
+    async def main(ofrak_context: OFRAKContext):
+    """
+    boilerplate_footer: str = r"""
+    if __name__ == "__main__":
+
+        # Use ANGR Backend
+        # import ofrak_angr
+        # import ofrak_capstone
+        # ofrak.discover(ofrak_capstone)
+        # ofrak.discover(ofrak_angr)
+
+        # Use Binary Ninja Backend
+        # import ofrak_binary_ninja
+        # import ofrak_capstone
+        # ofrak.discover(ofrak_capstone)
+        # ofrak.discover(ofrak_binary_ninja)
+
+        # Use Ghidra Backend
+        # import ofrak_ghidra
+        # ofrak.discover(ofrak_ghidra)
+
+        ofrak = OFRAK()
+        ofrak.run(main)
+    """
+
+
+class ScriptBuilder:
+    """
+    Builds and maintains runnable OFRAK scripts as sequences of actions, with each script tied to
+    a session.
+    """
+
+    def __init__(self):
+        self.root_cache: Dict[bytes, Resource] = {}
+        self.script_sessions: Dict[bytes, ScriptSession] = {}
+        self.selectable_indexes: List[ResourceIndexedAttribute] = [
+            FilesystemEntry.Name,
+            Data.Offset,
+        ]
+
+    async def _get_selector(self, resource: Resource) -> str:
+        root_resource = await self._get_root_resource(resource)
+        for ancestor in await resource.get_ancestors():
+            if ancestor.get_id() in self.script_sessions[root_resource.get_id()].variable_mapping:
+                break
+        attribute, attribute_value = await self._get_selectable_attribute(resource)
+        try:
+            result = await ancestor.get_only_child(
+                r_filter=ResourceFilter(
+                    tags=resource.get_most_specific_tags(),
+                    attribute_filters=[
+                        ResourceAttributeValueFilter(attribute=attribute, value=attribute_value)
+                    ],
+                )
+            )
+        except:
+            raise SelectableAttributesError(
+                f"Resource with ID {resource.get_id()} does not have a selectable attribute."
```

Comment:
> (this LoC will probably be too long so re-lint if you apply this suggestion)

Black strongly prefers not to add trailing `\` to break up long strings into multiple lines. So in this case, it doesn't automatically get split apart. Which is preferable: manually splitting up the LOC with `\`, splitting it up into 2 strings on different lines, or leaving it as-is?

Reply:
@marczalik Just go with whatever `black` wants to do, I don't think it ever uses backslashes to wrap strings and letting python auto-concat string literals on consecutive lines does the job.

@dannyp303 I see, that will be a good future extension. I suppose the reason why the chosen selector doesn't work is not of interest to the user in that context, so we don't really need to refine & define the reason. But, I still think this code should be refactored a little to form the loop I described. It seems logical, even for MVP, that if the resource has a selectable attribute that for some reason doesn't work, it will try the next one. And only if all selectable attributes either fail to be found, or don't form a successful query, do we then prompt the user.